ngrok/src/tunnel_ext.rs

```diff
@@ -0,0 +1,188 @@
+use std::{
+    convert::Infallible,
+    fmt,
+    io,
+    net::SocketAddr,
+};
+
+use async_trait::async_trait;
+use futures::stream::TryStreamExt;
+use hyper::{
+    server::conn::Http,
+    service::service_fn,
+    Body,
+    Response,
+    StatusCode,
+};
+use tokio::{
+    io::{
+        AsyncRead,
+        AsyncReadExt,
+        AsyncWrite,
+        AsyncWriteExt,
+    },
+    net::{
+        TcpStream,
+        ToSocketAddrs,
+    },
+    task::JoinHandle,
+};
+use tracing::{
+    debug,
+    field,
+    instrument,
+    trace,
+    warn,
+    Instrument,
+    Span,
+};
+
+use crate::{
+    prelude::*,
+    Conn,
+};
+
+impl<T> TunnelExt for T where T: Tunnel {}
+
+/// Extension methods auto-implemented for all tunnel types
+#[async_trait]
+pub trait TunnelExt: Tunnel {
+    /// Forward incoming tunnel connections to the provided TCP address.
+    #[instrument(level = "debug", skip_all, fields(local_addrs))]
+    async fn forward_tcp(&mut self, addr: impl ToSocketAddrs + Send) -> Result<(), io::Error> {
+        forward_conns(self, addr, |_, _| {}).await
+    }
+
+    /// Forward incoming tunnel connections to the provided TCP address.
+    ///
+    /// Provides slightly nicer errors when the backend is unavailable.
+    #[instrument(level = "debug", skip_all, fields(local_addrs))]
+    async fn forward_http(&mut self, addr: impl ToSocketAddrs + Send) -> Result<(), io::Error> {
+        forward_conns(self, addr, |e, c| drop(serve_gateway_error(e, c))).await
+    }
+}
+
+async fn forward_conns<T, A, F>(this: &mut T, addr: A, mut on_err: F) -> Result<(), io::Error>
+where
+    T: Tunnel + ?Sized,
+    A: ToSocketAddrs,
+    F: FnMut(io::Error, Conn),
+{
+    let span = Span::current();
+    let addrs = tokio::net::lookup_host(addr).await?.collect::<Vec<_>>();
+    span.record("local_addrs", field::debug(&addrs));
+    trace!("looked up local addrs");
+    loop {
+        trace!("waiting for new tunnel connection");
+        if !handle_one(this, addrs.as_slice(), &mut on_err).await? {
+            debug!("listener closed, exiting");
+            break;
+        }
+    }
+    Ok(())
+}
+
+async fn forward_bytes(
+    mut r: impl AsyncRead + Unpin,
+    mut w: impl AsyncWrite + Unpin,
+) -> Result<(), io::Error> {
+    let mut buf = vec![0u8; 1024];
+    loop {
+        let n = r.read(&mut buf).await?;
+        if n == 0 {
+            break;
+        }
+
+        w.write_all(&buf[0..n]).await?;
+    }
+    Ok(())
+}
+
+fn join_streams(
+    left: impl AsyncRead + AsyncWrite + Unpin + Send + 'static,
+    right: impl AsyncRead + AsyncWrite + Unpin + Send + 'static,
+) -> JoinHandle<()> {
+    let (t_rx, t_tx) = tokio::io::split(left);
+    let (l_rx, l_tx) = tokio::io::split(right);
```

Comment:
nit: would have expected (l_rx, l_tx) for left and (r_rx, r_tx) for right. maybe these were `tunnel` and `local` before they became `left` and `right`?

Reply:
heh, yep. I'll fix that.